internal/cache/base.go

```diff
@@ -0,0 +1,207 @@
+package cache
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	"fyne.io/fyne/v2"
+)
+
+var (
+	cacheDuration     = 1 * time.Minute
+	cleanTaskInterval = cacheDuration / 2
+
+	canvasRefreshCh = make(chan struct{}, 1)
+	expiredObjects  = make([]fyne.CanvasObject, 0, 50)
+	lastClean       time.Time
+)
+
+func init() {
+	if t, err := time.ParseDuration(os.Getenv("FYNE_CACHE")); err == nil {
+		cacheDuration = t
+		cleanTaskInterval = cacheDuration / 2
+	}
+}
+
+// CleanTask run cache clean task, it should be called on paint events.
+func CleanTask() {
+	now := time.Now()
+	// do not run clean task so fast
+	if now.Sub(lastClean) < 10*time.Second {
+		return
+	}
+	canvasRefresh := false
+	select {
+	case <-canvasRefreshCh:
+		canvasRefresh = true
+	default:
+		if now.Sub(lastClean) < cleanTaskInterval {
+			return
+		}
+	}
+	destroyExpiredSvgs(now)
+	if canvasRefresh {
+		// Destroy renderers on canvas refresh to avoid flickering screen.
+		destroyExpiredRenderers(now)
+		// canvases cache should be invalidated only on canvas refresh, otherwise there wouldn't
+		// be a way to recover them later
+		destroyExpiredCanvases(now)
+	}
+	lastClean = time.Now()
+}
+
+// ForceCleanFor forces a complete remove of all the objects that belong to the specified
+// canvas. Usually used to free all objects from a closing windows.
+func ForceCleanFor(canvas fyne.Canvas) {
+	deletingObjs := make([]fyne.CanvasObject, 0, 50)
+
+	// find all objects that belong to the specified canvas
+	canvasesLock.RLock()
+	for obj, cinfo := range canvases {
+		if cinfo.canvas == canvas {
+			deletingObjs = append(deletingObjs, obj)
+		}
+	}
+	canvasesLock.RUnlock()
+	if len(deletingObjs) == 0 {
+		return
+	}
+
+	// remove them from canvas cache
+	canvasesLock.Lock()
+	for _, dobj := range deletingObjs {
+		delete(canvases, dobj)
+	}
+	canvasesLock.Unlock()
+
+	// destroy their renderers and delete them from the renderer
+	// cache if they are widgets
+	renderersLock.Lock()
+	for _, dobj := range deletingObjs {
+		wid, ok := dobj.(fyne.Widget)
+		if !ok {
+			continue
+		}
+		winfo, ok := renderers[wid]
+		if !ok {
+			continue
+		}
+		winfo.renderer.Destroy()
+		delete(renderers, wid)
+	}
+	renderersLock.Unlock()
+}
+
+// NotifyCanvasRefresh notifies to the caches that a canvas refresh was triggered.
```

Comment:
This isn't really a notification - would `Request` work better than `Notify`?

Reply:
I don't think `Request` is the ideal replacement here because this function is not requesting a canvas refresh, it is just a way to notify to the cache that a `canvas.Refresh` was requested.

Reply:
I think it is unclear here the purpose of the API. A canvas callback would exist in the canvas code, not the cache.
Also this is not passing the canvas so it's a global flag - is that the right way to indicate a canvas is dirty?
With multiple windows we will be doing a lot of unnecessary work.

If you can outline what this is doing and who calls it for what purpose perhaps we come up with a better name and/or see if this is the right way to implement that requirement.

Reply:
> With multiple windows we will be doing a lot of unnecessary work.

cache.Clean is not calling when multiple windows are looping, so it won't do unnecessary work.

> If you can outline what this is doing and who calls it for what purpose perhaps we come up with a better name and/or see if this is the right way to implement that requirement.

Maybe I could remove that function by destroying the svgs on Canvas refresh too, previously it was needed for renderers but since renderers are going to be removed when canvas refresh, then this function should not be necessary. I will remove it, then :)

Reply:
done in 342712a